api/cmd/auth/auth.go

```diff
@@ -0,0 +1,95 @@
+package main
+
+import (
+	"fmt"
+
+	"net/http"
+	"os"
+
+	"github.com/gorilla/handlers"
+	"github.com/gorilla/mux"
+	"github.com/gorilla/sessions"
+	"github.com/markbates/goth"
+	"github.com/markbates/goth/gothic"
+	"github.com/markbates/goth/providers/bitbucket"
+	"github.com/markbates/goth/providers/github"
+	"github.com/markbates/goth/providers/gitlab"
+	"github.com/tektoncd/hub/api/pkg/app"
+	auth "github.com/tektoncd/hub/api/pkg/auth"
+)
+
+func main() {
+
+	key := ""            // Replace with your SESSION_SECRET or similar
+	maxAge := 86400 * 30 // 30 days
+	isProd := false      // Set to true when serving over https
+
+	store := sessions.NewCookieStore([]byte(key))
+	store.MaxAge(maxAge)
+	store.Options.Path = "/"
+	store.Options.HttpOnly = true // HttpOnly should always be enabled
+	store.Options.Secure = isProd
+
+	gothic.Store = store
+
+	api, err := app.FromEnv()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "FATAL: failed to initialise: %s", err)
+		os.Exit(1)
+	}
+
+	logger := api.Logger("main")
+	logger.Info("DB initialisation successful !!")
+
+	var AUTH_BASE_URL = os.Getenv("AUTH_BASE_URL")
+	var AUTH_URL = AUTH_BASE_URL + "/auth/%s/callback"
+
+	bitbucketAuth := auth.BitBucketProvider(AUTH_URL)
+	githubAuth := auth.GithubProvider(AUTH_URL)
+	gitlabAuth := auth.GitlabProvider(AUTH_URL)
+
+	goth.UseProviders(
+		github.NewCustomisedURL(
+			githubAuth.ClientId,
+			githubAuth.ClientSecret,
+			githubAuth.CallbackUrl,
+			githubAuth.AuthUrl,
+			githubAuth.TokenUrl,
+			githubAuth.ProfileUrl,
+			githubAuth.EmailUrl),
+
+		bitbucket.New(
+			bitbucketAuth.ClientId,
+			bitbucketAuth.ClientSecret,
+			bitbucketAuth.CallbackUrl,
+		),
+
+		gitlab.NewCustomisedURL(
+			gitlabAuth.ClientId,
+			gitlabAuth.ClientSecret,
+			gitlabAuth.CallbackUrl,
+			gitlabAuth.AuthUrl,
+			gitlabAuth.TokenUrl,
+			gitlabAuth.ProfileUrl,
+		),
+	)
+
+	r := mux.NewRouter()
+
+	authSvc := auth.New(api)
+
+	r.HandleFunc("/auth/login", authSvc.HubAuthenticate).Methods(http.MethodPost)
+
+	r.HandleFunc("/auth/{provider}/callback", authSvc.AuthCallBack)
+
+	r.HandleFunc("/auth/{provider}", auth.Authenticate)
+
+	logger.Info(fmt.Sprintf("listening on %s", AUTH_BASE_URL))
+
+	logger.Fatal(http.ListenAndServe(":4200",
```

Comment:
having this as separate dir, does it means it will have a separate image?

Reply:
if it is possible we should try to run this on different port but with the existing api server

Reply:
running this server as a thread

Reply:
maintaining different images not a good idea if we can package in one

Reply:
Yes it will have a separate image

I think the ultimate aim of having a separate auth server is it's reusability and transparency from api server, running it in existing api server with different port might break this. Also if someone just wants the auth server, but building the image would also have the api in it, which would  be of no use for the user

Also do you mean adding it in existing server means designing the api endpoints for auth using GOA, 
If yes then for auth we have used Gorilla mux, the reason behind using gorilla was Goth library which we are using for authentication requires http.ResponseWriter and http.Request, which were not able to access using GOA